internal/service/event_listener.go

```diff
@@ -0,0 +1,116 @@
+	// coming from systemd
+	servicePostfixLength = 8
+
+	EventStarted EventType = "started"
+	EventStopped EventType = "stopped"
+
+	// To avoid confusion, we match the action verb tense coming from systemd sub state, even though it is not consistent
+	// The only addition to the list is the "removed" case, which we use when a service is removed
+	// from the system and we receive an empty unitStatus object.
+
+	//The service is running
+	running unitSubState = "running"
+	//The service is starting, probably due to a restart.
+	//In this case we have to notify the observers that the service is not yet up and is loading. A subsequent 'running' event will proceed.
+	start unitSubState = "start"
+	//The service has been stopped, due to the container being killed. Systemd will restart the service.
+	stop unitSubState = "stop"
+	//The service is dead, due to the container being stopped for instance.
+	dead unitSubState = "dead"
+	//The service failed to start
+	failed unitSubState = "failed"
+	//The service has been removed, no unit state is returned from systemd
+	removed unitSubState = "removed"
+)
+
+type unitSubState string
+
+type EventType string
+
+type Event struct {
+	WorkloadName string
+	Type         EventType
+}
+
+type EventListener struct {
+	observerCh chan *Event
+	set        *dbus.SubscriptionSet
+	dbusCh     <-chan map[string]*dbus.UnitStatus
+	dbusErrCh  <-chan error
+}
+
+func NewEventListener() *EventListener {
+	return &EventListener{}
+}
+
+func (e *EventListener) Connect() (chan *Event, error) {
+	conn, err := newDbusConnection(UserBus)
+	if err != nil {
+		return nil, err
+	}
+	e.set = conn.NewSubscriptionSet()
+	e.observerCh = make(chan *Event)
+	return e.observerCh, nil
+}
+
+func (e *EventListener) Listen() chan *Event {
+	e.dbusCh, e.dbusErrCh = e.set.Subscribe()
+	for {
+		select {
+		case msg := <-e.dbusCh:
+			for name, unit := range msg {
+				log.Debugf("Captured event for %s: %v+", name, unit)
+				n := extractWorkloadName(name)
+				state := translateUnitSubStatus(unit)
+				log.Debugf("Systemd service for workload %s transitioned to sub state %s\n", n, state)
+				switch state {
+				case running:
+					e.observerCh <- &Event{WorkloadName: n, Type: EventStarted}
+				case removed, stop, dead, failed, start:
+					e.observerCh <- &Event{WorkloadName: n, Type: EventStopped}
+				default:
+					log.Debugf("Ignoring unit sub state for service %s: %s", name, unit.SubState)
+				}
+			}
+		case msg := <-e.dbusErrCh:
+			log.Errorf("Error while parsing dbus event: %v", msg)
+		}
+	}
+
+}
+
+func translateUnitSubStatus(unit *dbus.UnitStatus) unitSubState {
+	if unit == nil {
+		return removed
+	}
+	return unitSubState(unit.SubState)
+}
+
+func (e *EventListener) Add(workloadName string) {
+	name := fmt.Sprintf("%s.service", workloadName)
+	log.Debugf("Adding service for events %s", name)
+	if !e.set.Contains(name) {
```

Comment:
it's a safeguard just in case the service was already part of the filter. In theory this should never happen based on the nominal flow. If a service is replaced it will be first removed and then readded by the events.

Reply:
Ended up removing it with the new logic to watch for file system events. The case where the service could be added twice is not possible anymore.